crates/runtime/src/traphandlers.rs

```diff
@@ -783,3 +799,40 @@ fn setup_unix_sigaltstack() -> Result<(), Trap> {
         }
     }
 }
+
+/// Perform a program exit by unwinding the stack to the point where wasm
+/// as most recently entered, carrying an exit status value.
+///
+/// This is implemented in `wasmtime_runtime` rather than with the rest of the WASI
+/// functions as it's essentially an unwinding operation.
+///
+/// # Safety
+///
+/// Only safe to call when wasm code is on the stack, aka `catch_traps` must
+/// have been previously called. Additionally no Rust destructors can be on the
+/// stack. They will be skipped and not executed.
+pub unsafe extern "C" fn wasi_proc_exit(
+    _vmctx: *mut VMContext,
+    _caller_vmctx: *mut VMContext,
+    status: i32,
+) -> ! {
+    // Check that the status is within WASI's range.
+    if status >= 0 && status < 126 {
+        tls::with(|info| info.unwrap().unwind_with(UnwindReason::Exit { status }))
+    } else {
+        raise_user_trap(InvalidWASIExitStatus {}.into())
```

Comment:
Given that we only really have "trusted" code calling this, could this be a panic or something more general? This seems a bit low-level to be dealing with WASI-specific codes and things.